sys/include/net/gcoap/dns.h

```diff
@@ -0,0 +1,200 @@
+#define CONFIG_GCOAP_DNS_CREDS_MAX          1U
+#else
+#define CONFIG_GCOAP_DNS_CREDS_MAX          0U
+#endif
+#endif
+
+/**
+ * @brief   size of the buffer used to build a CoAP request
+ */
+#ifndef CONFIG_GCOAP_DNS_PDU_BUF_SIZE
+#define CONFIG_GCOAP_DNS_PDU_BUF_SIZE       128U
+#endif
+
+/**
+ * @brief   block-size used for requests (both Block2 control and Block1)
+ *
+ * Should be less than @ref CONFIG_GCOAP_DNS_PDU_BUF_SIZE or
+ * @ref CONFIG_GCOAP_PDU_BUF_SIZE and must be a power
+ * of 2
+ */
+#ifndef CONFIG_GCOAP_DNS_BLOCK_SIZE
+#define CONFIG_GCOAP_DNS_BLOCK_SIZE         64U
+#endif
+/** @} */
+
+/**
+ * @brief   Query a domain name via CoAP synchronously
+ *
+ * @note    Only available with module `gcoap_dns_async` compiled in.
+ *
+ * @param[in] domain_name   A '\0'-terminated domain name. Must not be NULL.
+ * @param[out] addr_out     The resolved address. Must not be NULL.
+ * @param[in] family        The desired address family for @p addr_out.
+ *                          @ref AF_UNSPEC for any address family (an IPv6
+ *                          address will take preference over an IPv4 address).
+ *
+ * @return  length of @p addr_out in bytes on success
+ * @return  -EAFNOSUPPORT, if the hostname of the URI resolves to an unknown address family.
+ * @return  -EBADMSG, when receiving erroneous response or response containing
+ *          an error code.
+ * @return  -ECONNABORTED, if CoAP request cannot be sent.
+ * @return  -ECONNREFUSED, if no URI is set for the client (see @ref gcoap_dns_server_uri_set()).
+ * @return  -EDESTADDRREQ, if CoAP response was received from an unexpected
+ *          remote.
+ * @return  -EHOSTUNREACH, if the hostname of the URI can not be resolved
+ * @return  -ENOBUFS, if there was not enough buffer space for the request.
+ * @return  -ENOBUFS, if length of received CoAP body is greater than
+ *          @ref CONFIG_DNS_MSG_LEN.
+ * @return  -ENOENT, if Zone-ID of the URI can not be found locally.
+ * @return  -ENOMSG, if CoAP response did not contain a DNS response.
+ * @return  -ENOTRECOVERABLE, on gCoAP-internal error.
+ * @return  -ENOTSUP, if credential can not be added for to client.
+ * @return  -ENOTSUP, if module `gcoap_dns_async` is not compiled in.
+ * @return  -ETIMEDOUT, if CoAP request timed out.
+ */
+int gcoap_dns_query(const char *domain_name, void *addr_out, int family);
+
+/**
+ * @brief   Sets and checks a URI
+ *
+ * @param[in] uri   A URI. May be NULL to remove the URI.
+ *
+ * @return  length of @p uri on success (0 when URI is removed).
+ * @return  -EINVAL, if @p uri is not a valid URI for DNS over CoAP.
+ * @return  -ENOBUFS, if the client implementation has not enough buffer space
+ *          to process the URI.
+ */
+int gcoap_dns_server_uri_set(const char *uri);
+
+/**
+ * @brief   Gets the URI
+ *
+ * This function is not thread-safe.
+ *
+ * @return  The current URI for the client.
+ */
+const char *gcoap_dns_server_uri_get(void);
```

Comment:
Why do we need this public? That's hard to keep up with this signature if (or when) we move the stored URI to a differently structured form (eg. preparsed, cf. https://github.com/RIOT-OS/RIOT/issues/13827) to reduce parsing overhead on each DNS request.

Reply:
Mainly so a user can check if a server URI is already set. Would making this having an out parameter and numerical error code as return value make this easier for you.

Reply:
Something like `bool gcoap_dns_server_uri_is_set(void)` would work with any kind of processing. A `int gcoap_dns_server_uri_get(char *uri, size_t uri_len);` would work in theory, but entails a lot of string parsing, and if "did I already" is the only purpose, is_set suffices. (Or really, so does just setting it again).

Reply:
> and if "did I already" is the only purpose, is_set suffices. (Or really, so does just setting it again).

Make that "did someone already" ;-) For DNS applications I regularly use the equivalent to check if the DNS server was set via RA.

Reply:
> A `int gcoap_dns_server_uri_get(char *uri, size_t uri_len);` would work in theory, but entails a lot of string parsing,

Does it? Just for testing purposes I would prefer to be able to check what the client things it is using.

Reply:
Adapted for both now `is_set` when only a simple check is required, get to get the actual URI (e.g. for testing).

Reply:
Formatting a preprocessed list of CoAP options into a URI is something I'd like to avoid pulling in from system components; it's at least hex-formatting IPv6 addresses and decimal-formatting v4 ones (why would a device w/o direct user input do either?), and can get a lot more complex when host names are involved (percent encoding, possibly punycode although we probably don't handle that correctly on the input side).

Having the _is_set here at least ensures that if we deprecate (or API-change _get) at any point, it only hits those who really wanted the URI (and not those just checking for whether things are configured).